### PR TITLE
Enable Integration tests for MacOS

### LINKED ISSUE #1105: [Request]: Integration tests for macOs
### Plugin

all

### Use case

At the moment integration_test run on Android and iOS emulator/simulator but not on desktop

### Proposal

Enable the integration_test to run and pass on a macOs "macos-latest" machine.

Also, the tests will need to be adapted to macOs.

### PR DESCRIPTION
**Pull request by werainkhatri**
## Description

Adds workflow steps to run integration tests for MacOS for plugins:
- battery_plus
- connectivity_plus
- device_info_plus
- network_info_plus
- package_info_plus

Also, tweaked a few tests to adapt for MacOS

These packages / APIs aren't implemented in MacOS yet, so haven't added them.
- `sensors_plus`
- `share_plus`
- `battery_plus`'s `isInBatterySaveMode`

## Related Issues

Fixes #1105

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [ ] I updated the version in `pubspec.yaml` and `CHANGELOG.md`.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/fluttercommunity/plus_plugins/blob/main/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

### CODE REVIEW

**Comment by miquelbeltran** [comment]
macos_integration_test for connectivity_plus failed but the job appears in green,

See the logs:

```
02:10 +0: loading /Users/runner/work/plus_plugins/plus_plugins/packages/connectivity_plus/connectivity_plus/example/integration_test/connectivity_plus_test.dart                                       
02:10 +0: Connectivity test driver (setUpAll)                                                                                                                                                          
02:11 +0: Connectivity test driver (setUpAll)                                                                                                                                                          
02:11 +0: Connectivity test driver test connectivity result                                                                                                                                            
02:11 +1: Connectivity test driver test connectivity result                                                                                                                                            
02:11 +1: Connectivity test driver connectivity on macos should be wifi                                                                                                                                
02:11 +1: Connectivity test driver connectivity on macos should be wifi                                                                                                                                
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following TestFailure was thrown running a test:
Expected: ConnectivityResult:<ConnectivityResult.wifi>
  Actual: ConnectivityResult:<ConnectivityResult.ethernet>

When the exception was thrown, this was the stack:
#4      main.<anonymous closure>.<anonymous closure> (file:///Users/runner/work/plus_plugins/plus_plugins/packages/connectivity_plus/connectivity_plus/example/integration_test/connectivity_plus_test.dart:40:7)
<asynchronous suspension>
<asynchronous suspension>
(elided one frame from package:stack_trace)

This was caught by the test expectation on the following line:
  file:///Users/runner/work/plus_plugins/plus_plugins/packages/connectivity_plus/connectivity_plus/example/integration_test/connectivity_plus_test.dart line 40
The test description was:
  connectivity on macos should be wifi
════════════════════════════════════════════════════════════════════════════════════════════════════

02:11 +1 -1: Connectivity test driver connectivity on macos should be wifi [E]                                                                                                                         
  Test failed. See exception logs above.
  The test description was: connectivity on macos should be wifi
```

File: packages/battery_plus/battery_plus/example/integration_test/battery_plus_test.dart
Comment (miquelbeltran) [inline packages/battery_plus/battery_plus/example/integration_test/battery_plus_test.dart:27]: Could you document why this tests needs to be skipped for MacOS?

Maybe it is better to write `skip: Platform.isMacOs` instead.

File: packages/connectivity_plus/connectivity_plus/example/integration_test/connectivity_plus_test.dart
Comment (miquelbeltran) [inline packages/connectivity_plus/connectivity_plus/example/integration_test/connectivity_plus_test.dart:36]: I'd rather not do this, and instead just name it `Connectivity should be wifi`

File: packages/connectivity_plus/connectivity_plus/example/integration_test/connectivity_plus_test.dart
Comment (miquelbeltran) [inline packages/connectivity_plus/connectivity_plus/example/integration_test/connectivity_plus_test.dart:41]: As we saw on the test run results on GitHub Actions, MacOS expects ConnectivityResult ethernet.

File: packages/network_info_plus/network_info_plus/example/integration_test/network_info_plus_test.dart
Comment (miquelbeltran) [inline packages/network_info_plus/network_info_plus/example/integration_test/network_info_plus_test.dart:24]: this line should be removed, otherwise it won't run on macos

File: packages/connectivity_plus/connectivity_plus/example/integration_test/connectivity_plus_test.dart
Comment (miquelbeltran) [reply packages/connectivity_plus/connectivity_plus/example/integration_test/connectivity_plus_test.dart:36]: Or maybe just write one test per platform, it is ok if tests are duplicated IMO

**Comment by miquelbeltran** [comment]
Can you add share_plus as well?

**Comment by werainkhatri** [comment]
> Can you add share_plus as well?

I was just going to document that. I did attempt writing a test, but `Share.share` (and `Share.shareWithResult` API(s) for MacOS opens up a native list dialog, and the function does not complete unless an item is selected from the dialog or it is dismissed. I attempted dismissing it using `tester.tap`, but soon realizsed that the tap is created and consumed by flutter instead of the operating system. Not sure how to focus w.r.t the operating system. Will need to dive deeper into understanding that.

File: packages/battery_plus/battery_plus/example/integration_test/battery_plus_test.dart
Comment (werainkhatri) [reply packages/battery_plus/battery_plus/example/integration_test/battery_plus_test.dart:27]: i'll document this in the getter itself. fyi: this feature is not implemented in MacOS yet.

File: packages/connectivity_plus/connectivity_plus/example/integration_test/connectivity_plus_test.dart
Comment (werainkhatri) [reply packages/connectivity_plus/connectivity_plus/example/integration_test/connectivity_plus_test.dart:36]: oh okay sure.

File: packages/connectivity_plus/connectivity_plus/example/integration_test/connectivity_plus_test.dart
Comment (werainkhatri) [reply packages/connectivity_plus/connectivity_plus/example/integration_test/connectivity_plus_test.dart:41]: hmm weird. i pre-ran these tests on my fork and it worked just fine. any idea?

workflow job in a pr on my fork: https://github.com/werainkhatri/plus_plugins/actions/runs/3164967700/jobs/5153647025

File: packages/network_info_plus/network_info_plus/example/integration_test/network_info_plus_test.dart
Comment (werainkhatri) [reply packages/network_info_plus/network_info_plus/example/integration_test/network_info_plus_test.dart:24]: oh my bad. thanks for catching that! didn't expect that to be there 😅

File: packages/connectivity_plus/connectivity_plus/example/integration_test/connectivity_plus_test.dart
Comment (werainkhatri) [reply packages/connectivity_plus/connectivity_plus/example/integration_test/connectivity_plus_test.dart:41]: well, on my side the check seems to have passed. could you take another look?

**Comment by werainkhatri** [comment]
> Can you add share_plus as well?

if you just want me to add it to the actions list (i can do the same for sensors_plus), i'll do that. but there are no tests for it. let me know your thoughts.

File: packages/network_info_plus/network_info_plus/example/integration_test/network_info_plus_test.dart
Comment (werainkhatri) [reply packages/network_info_plus/network_info_plus/example/integration_test/network_info_plus_test.dart]: turns out, this is only implemented in IOS (as the comment said). reverted. sorry bout that.

File: packages/connectivity_plus/connectivity_plus/example/integration_test/connectivity_plus_test.dart
Comment (miquelbeltran) [reply packages/connectivity_plus/connectivity_plus/example/integration_test/connectivity_plus_test.dart]: Maybe your laptop was connected to Wi-Fi but the GitHub Action images run on emulated Ethernet :)

**Comment by miquelbeltran** [comment]
> > Can you add share_plus as well?
> 
> I was just going to document that. I did attempt writing a test, but `Share.share` (and `Share.shareWithResult`) API(s) for MacOS open up a native list dialog, and the function does not complete unless an item is selected from the dialog or it is dismissed. I attempted dismissing it using `tester.tap`, but soon realised that the tap is created and consumed by flutter instead of the operating system. Not sure how to focus w.r.t the operating system. Will need to dive deeper into understanding that.

I think it will be fine just by making the call and verifying that no crash/exception happen. I think we faced the same issue on iOS as well.

File: packages/connectivity_plus/connectivity_plus/example/integration_test/connectivity_plus_test.dart
Comment (werainkhatri) [reply packages/connectivity_plus/connectivity_plus/example/integration_test/connectivity_plus_test.dart]: by "on my side", i meant the check on github states that it succeeded.

workflow job on this pr: https://github.com/fluttercommunity/plus_plugins/actions/runs/3165195385/jobs/5154030407

**Comment by werainkhatri** [comment]
why aren't the checks still pending?

**Comment by vbuberen** [comment]
> why aren't the checks still pending?

Because of other PRs occupying all the available resources. Had to manually cancel some runs for another PR to unblock us.

File: packages/battery_plus/battery_plus/example/integration_test/battery_plus_test.dart
Comment (miquelbeltran) [inline packages/battery_plus/battery_plus/example/integration_test/battery_plus_test.dart:18]: Seems that `Can get battery level` won't work on iOS Simulators, so this needs to be changed accordingly to pass the tests

**Comment by miquelbeltran** [comment]
A couple of things that should be improved before approval:

- on battery_plus, all tests are skipped for MacOS:

```
01:38 +0 ~2: All tests skipped. 
```

- on connectivity_plus, tests are failing but the job appears in green

```
When the exception was thrown, this was the stack:
#4      main.<anonymous closure>.<anonymous closure> (file:///Users/runner/work/plus_plugins/plus_plugins/packages/connectivity_plus/connectivity_plus/example/integration_test/connectivity_plus_test.dart:39:7)
<asynchronous suspension>
<asynchronous suspension>
(elided one frame from package:stack_trace)

This was caught by the test expectation on the following line:
  file:///Users/runner/work/plus_plugins/plus_plugins/packages/connectivity_plus/connectivity_plus/example/integration_test/connectivity_plus_test.dart line 39
The test description was:
  connectivity on MacOS should be wifi
════════════════════════════════════════════════════════════════════════════════════════════════════

03:23 +1 ~1 -1: Connectivity test driver connectivity on MacOS should be wifi [E]                                                                                                                      
  Test failed. See exception logs above.
  The test description was: connectivity on MacOS should be wifi
  

To run this test again: /Users/runner/work/plus_plugins/plus_plugins/_flutter/bin/cache/dart-sdk/bin/dart test /Users/runner/work/plus_plugins/plus_plugins/packages/connectivity_plus/connectivity_plus/example/integration_test/connectivity_plus_test.dart -p vm --plain-name 'Connectivity test driver connectivity on MacOS should be wifi'

03:23 +1 ~1 -1: Connectivity test driver (tearDownAll)                                                                                                                                                 
03:23 +1 ~1 -1: (tearDownAll)                                                                                                                                                                          
03:23 +1 ~1 -1: Some tests failed.                                                 
```

The others seem to be fine!

**Comment by werainkhatri** [comment]
"Can get battery level" for battery_plus on MacOS is silently failing, not sure why. the error is not logged. checking...

```
01:39 +0: Can get battery level                                                                                                                                                                        
01:40 +0: Can get battery level                                                                                                                                                                        
01:40 +0 ~1: Can get battery level                                                                                                                                                                     
01:40 +0 ~1: Can get battery level - did not complete [E]                                                                                                                                              

01:40 +0 ~1: (tearDownAll) - did not complete [E]                                                                                                                                                      

01:40 +0 ~1: Some tests failed.
```

**Comment by miquelbeltran** [comment]
> "Can get battery level" for battery_plus on MacOS is silently failing, not sure why. the error is not logged. checking...
> 
> ```
> 01:39 +0: Can get battery level                                                                                                                                                                        
> 01:40 +0: Can get battery level                                                                                                                                                                        
> 01:40 +0 ~1: Can get battery level                                                                                                                                                                     
> 01:40 +0 ~1: Can get battery level - did not complete [E]                                                                                                                                              
> 
> 01:40 +0 ~1: (tearDownAll) - did not complete [E]                                                                                                                                                      
> 
> 01:40 +0 ~1: Some tests failed.
> ```

You could create a new test for batteryState, for example

**Comment by werainkhatri** [comment]
> If you find that no test completes due to weirdness of the macos vm, then skip the job completely and add some documentation explaining why.

ya that seems to be the case. i'm trying one last time (on my fork PR). will update accordingly.

**Review by miquelbeltran — APPROVED** [review]

**Comment by miquelbeltran** [comment]
Thanks! PR should be fine now, I will wait until all jobs complete green to merge

**Comment by werainkhatri** [comment]
thank you for your patience and i apologize for the long back and forth. integration testing and melos were new territory for me.

i had a query (will raise an issue consequently), why can checks fail silently? is it an issue in the way melos is configured?

**Comment by werainkhatri** [comment]
@miquelbeltran could you take a look at the failing check? looks irrelevant / flaky to me.

**Comment by miquelbeltran** [comment]
I'm on my phone, but looks like just pub get failed (network issue?). Nothing to worry about.

**Pull request merged**
